tooling/miniutil/src/fmt/expr.rs

```diff
@@ -1,21 +1,58 @@
 use super::*;
 
-pub(super) fn fmt_place_expr(p: PlaceExpr, comptypes: &mut Vec<CompType>) -> String {
+// A formatted expression.
+// This type is used to insert the minimal required amount of parens to make expressions unambiguous, without implementing an operator priority.
+pub(super) enum FmtExpr {
+    // An expression that might be ambiguous in certain contexts.
+    // For example `a + b` is ambiguous in the context `a + b * c`: It might mean `(a + b) * c` or `a + (b * c)`
+    NonAtomic(String),
+
+    // An expression that is never ambiguous.
+    // For example `2`, `(a + b)` or `load(_1)`.
+    Atomic(String),
+}
+
+impl FmtExpr {
+    // Returns the contents of this FmtExpr as-is, without wrapping in `(`, `)`.
+    // Use this function in unambiguous contexts like `foo(_)`.
+    pub(super) fn to_string(self) -> String {
+        match self {
+            FmtExpr::NonAtomic(s) => s,
+            FmtExpr::Atomic(s) => s,
+        }
+    }
+
+    // Wraps the expression in `(`, `)` if necessary.
+    // Use this function in ambiguous contexts like `a + _`.
+    pub(super) fn to_atomic_string(self) -> String {
+        match self {
+            // This adds parens around non-atomic expressions to make them atomic.
+            FmtExpr::NonAtomic(s) => format!("({s})"),
+            FmtExpr::Atomic(s) => s,
+        }
+    }
+}
+
+pub(super) fn fmt_place_expr(p: PlaceExpr, comptypes: &mut Vec<CompType>) -> FmtExpr {
     match p {
-        PlaceExpr::Local(l) => fmt_local_name(l),
+        PlaceExpr::Local(l) => FmtExpr::Atomic(fmt_local_name(l)),
         PlaceExpr::Deref { operand, ptype } => {
-            let ptype = fmt_ptype(ptype, comptypes);
-            let expr = fmt_value_expr(operand.extract(), comptypes);
-            format!("deref<{ptype}>({expr})")
+            let ptype = fmt_ptype(ptype, comptypes).to_string();
+            let expr = fmt_value_expr(operand.extract(), comptypes).to_string();
+            FmtExpr::Atomic(format!("deref<{ptype}>({expr})"))
         }
         PlaceExpr::Field { root, field } => {
-            let root = fmt_place_expr(root.extract(), comptypes);
-            format!("{root}.{field}")
+            let root = fmt_place_expr(root.extract(), comptypes).to_atomic_string();
+            // You might think that `&raw foo.bar` is ambiguous and hence `foo.bar` is non-atomic.
+            // But we define `&raw foo.bar` to always mean `&raw (foo.bar)` and not `(&raw foo).bar`.
+            // Hence this can be atomic.
```

Comment:
```suggestion
            let root = fmt_place_expr(root.extract(), comptypes).to_atomic_string();
            // `&raw foo.bar` in Rust unambiguously means `&raw (foo.bar)`, and there is
            // no other context we have to worry about. Hence this can be atomic.
```
We don't get to define that, that's a Rust thing.^^